File: 110.balanced-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
	public boolean isBalanced(TreeNode root) {
		if (root == null) return true;
		return isBalancedHelper(root) != -1;        
	}

	public int isBalancedHelper(TreeNode node){
		if(node == null) return 0;
		int leftHeight = isBalancedHelper(node.left);
		if(leftHeight == -1) return -1;
		int rightHeight = isBalancedHelper(node.right);
		if(rightHeight == -1 ) return -1;
		if (leftHeight - rightHeight > 1 || leftHeight - rightHeight < -1){
			return -1;
		}
		return 1+Math.max(leftHeight,rightHeight); 
	}
}

File: 110.balanced-binary-tree.js
/*
 * @lc app=leetcode id=110 lang=javascript
 *
 * [110] Balanced Binary Tree
 */
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
// var isBalanced = function (root) {
//   if(!root) {
//     return true;
//   }


//   let left = isBalancedHelper(root.left);
//   let right = isBalancedHelper(root.right);
//   return Math.abs(left - right) > 1 ? false : true;
// };

// isBalancedHelper = (root) => {
//   if (!root) {
//     return 0;
//   }
//   let left = isBalancedHelper(root.left);
//   let right = isBalancedHelper(root.right);
//   return Math.max(left, right) + 1;
// }


var isBalanced = function (root) {
  return isBalancedHelper(root) != -1;
};

isBalancedHelper = (root) => {
  if (!root) {
    return 0;
  }
  let left = isBalancedHelper(root.left);
  let right = isBalancedHelper(root.right);
  if(left == -1 || right == -1  ||  Math.abs(left - right) > 1) {
    return -1;
  } 
  return Math.max(left, right) + 1;
}

